src/components/marker.js

```diff
@@ -53,14 +53,25 @@ export default class Marker extends BaseControl {
     const {className, longitude, latitude, offsetLeft, offsetTop} = this.props;
 
     const [x, y] = this.context.viewport.project([longitude, latitude]);
-    const containerStyle = {
-      position: 'absolute',
+
+    const {left, top} = {
       left: x + offsetLeft,
       top: y + offsetTop
     };
 
+    const isVisible = !(top < 0 || left < 0);
+
+    let containerStyle = {};
+
+    if (isVisible) {
+      containerStyle = {
+        transform: `translate3d(${left}px, ${top}px, 0)`
+      };
+    }
+
     return createElement('div', {
-      className: `mapboxgl-marker ${className}`,
+      // isVisible toggles the .visible to show/hide the element
+      className: `mapboxgl-marker ${className} ${isVisible ? 'visible' : ''}`,
```

Comment:
I can see that avoid rendering out-of-view markers provide perf benefit. Though I'm not convinced about the `visible` class. It adds an additional catch for the user to implement, and it doesn't account for the size of the marker.  @abmai

Reply:
Am I need to put the marker sizes check. Or it would be better to roll it back?

Reply:
@rshtg We don't ask the user to specify the size of each marker today. Dynamically querying their sizes may cause perf problems. Shall we remove it for now to land the transform change?

Reply:
I think we just can keep only transformation without `visible` toggler. So the profit should be affected in any case